src/stroke.rs

```diff
@@ -0,0 +1,630 @@
+    Round,
+}
+
+/// Describes the visual style of a stroke.
+#[derive(Clone, Debug)]
+pub struct Stroke {
+    /// Width of the stroke.
+    pub width: f64,
+    /// Style for connecting segments of the stroke.
+    pub join: Join,
+    /// Limit for miter joins.
+    pub miter_limit: f64,
+    /// Style for capping the beginning of an open subpath.
+    pub start_cap: Cap,
+    /// Style for capping the end of an open subpath.
+    pub end_cap: Cap,
+    /// Lengths of dashes in alternating on/off order.
+    pub dash_pattern: Dashes,
+    /// Offset of the first dash.
+    pub dash_offset: f64,
+    /// True if the stroke width should be affected by the scale of a
+    /// transform.
+    ///
+    /// Discussion question: does this make sense here?
+    pub scale: bool,
+}
+
+impl Default for Stroke {
+    fn default() -> Self {
+        Self {
+            width: 1.0,
+            join: Join::Round,
+            miter_limit: 4.0,
+            start_cap: Cap::Round,
+            end_cap: Cap::Round,
+            dash_pattern: Default::default(),
+            dash_offset: 0.0,
+            scale: true,
+        }
+    }
+}
+
+impl Stroke {
+    /// Creates a new stroke with the specified width.
+    pub fn new(width: f64) -> Self {
+        Self {
+            width,
+            ..Default::default()
+        }
+    }
+
+    /// Builder method for setting the join style.
+    pub fn with_join(mut self, join: Join) -> Self {
+        self.join = join;
+        self
+    }
+
+    /// Builder method for setting the limit for miter joins.
+    pub fn with_miter_limit(mut self, limit: f64) -> Self {
+        self.miter_limit = limit;
+        self
+    }
+
+    /// Builder method for setting the cap style for the start of the stroke.
+    pub fn with_start_cap(mut self, cap: Cap) -> Self {
+        self.start_cap = cap;
+        self
+    }
+
+    /// Builder method for setting the cap style for the end of the stroke.
+    pub fn with_end_cap(mut self, cap: Cap) -> Self {
+        self.end_cap = cap;
+        self
+    }
+
+    /// Builder method for setting the cap style.
+    pub fn with_caps(mut self, cap: Cap) -> Self {
+        self.start_cap = cap;
+        self.end_cap = cap;
+        self
+    }
+
+    /// Builder method for setting the dashing parameters.
+    pub fn with_dashes<P>(mut self, offset: f64, pattern: P) -> Self
+    where
+        P: IntoIterator,
+        P::Item: Borrow<f64>,
+    {
+        self.dash_offset = offset;
+        self.dash_pattern.clear();
+        self.dash_pattern
+            .extend(pattern.into_iter().map(|dash| *dash.borrow()));
+        self
+    }
+
+    /// Builder method for setting whether or not the stroke should be affected
+    /// by the scale of any applied transform.
+    pub fn with_scale(mut self, yes: bool) -> Self {
+        self.scale = yes;
+        self
+    }
+}
+
+/// Collection of values representing lengths in a dash pattern.
+pub type Dashes = SmallVec<[f64; 4]>;
```

Comment:
It may be better to do three dashes by default, since on 64 bit platforms there would be no wasted stack space compared to a `Vec`.